### ISSUE TITLE
'ZERO WIDTH NO-BREAK SPACE' causes issues with first column value lookup

### ISSUE THREAD

**Issue by apoorvam** [getgauge/gauge#509]
_From @ryanrust on August 17, 2016 18:0_

### Expected behavior
If a .csv file that starts with a ['ZERO WIDTH NO-BREAK SPACE'](http://www.fileformat.info/info/unicode/char/feff/index.htm) is passed to a Step implementation using a Csv parameter, looking up row values with **TableRow.GetCell()** should work normally.

### Actual behavior
The character is included as part of the key for the header for the first column in the file and it is impossible to lookup the column values using **TableRow.GetCell()**. Specifically, this occurs because the whitespace character is included as part of the string in the header key collection so lookup by that key always fails.

### Steps to reproduce
1. Create a Step that passes a .csv file starting with a 'ZERO WIDTH NO-BREAK SPACE' to an implementation with a Csv parameter. (Note that the easiest way I've found to consistently create a file with this character at the beginning is using the "Add->New Item" context menu in Visual Studio. Choose "Text File" and change the Name to "<filename>.csv".)

2. Attempt to retrieve a value from the first column using the **TableRow.GetCell()** method.

[Gist](https://gist.github.com/ryanrust/7b7bf76244f932fcee62215bb90e1b85)

### Gauge version
```
Gauge version: 0.5.0

Plugins
-------
csharp (0.8.0)
html-report (2.1.1)
java (0.4.1)
xml-report (0.1.3)
```

_Copied from original issue: getgauge/gauge-csharp#91_

**Comment by apoorvam** [comment]
_From @sguptatw on August 18, 2016 16:38_

@ryanrust : Can you please share your use case with us for having this specific csv file name. Would renaming the file resolve the issue you are facing?

**Comment by apoorvam** [comment]
_From @ryanrust on August 18, 2016 16:39_

I linked an example Use Case in the Gist below the "Steps to Reproduce". It's not an issue with the filename but rather an issue with the contents of the file.

**Comment by apoorvam** [comment]
_From @sriv on August 19, 2016 6:13_

I can replicate this. With `Gauge.CSharp.Lib` 0.6.1 released, one can print the raw table in Markdown format. Using which I was able to change @ryanrust's implementation to 

```
        [Step("Pass file <contents>.")]
        public void PassFile(Table contents)
        {
            Console.WriteLine(contents);
            // Should be "TestVal1" but is null.
            Debug.WriteLine("Column1: " + contents.GetTableRows().First().GetCell("Column1"));
            Console.WriteLine("Column1: " + contents.GetTableRows().First().GetCell("Column1"));

            // Other columns work properly.
            Debug.WriteLine("Column2: " + contents.GetTableRows().First().GetCell("Column2"));
            Debug.WriteLine("Column3: " + contents.GetTableRows().First().GetCell("Column3"));
            Console.WriteLine("Column2: " + contents.GetTableRows().First().GetCell("Column2"));
            Console.WriteLine("Column3: " + contents.GetTableRows().First().GetCell("Column3"));
        }

```

Running `gauge specs` yields

```
Gauge.CSharp.Runner.GaugeProjectBuilder Building Project: C:/Users/srikantv/AppData/Local/Temp/foo/foo.csproj
Gauge.CSharp.Runner.GaugeProjectBuilder Success
Gauge.CSharp.Runner.SandboxBuilder Creating a Sandbox in: C:\Users\srikantv\AppData\Local\Temp\foo\gauge-bin
Sandbox Loaded Instance Manager of Type:Gauge.CSharp.Lib.DefaultClassInstanceManager
# Example
  ## Pass a file        Sandbox foo.StepImplementation
|?Column1|Column2 |Column3 |
|--------|--------|--------|
|Testval1|Testval2|Testval3|
Column1:
Column2: Testval2
Column3: Testval3
 P

Successfully generated html-report to => C:\Users\srikantv\AppData\Local\Temp\foo\reports\html-report
Specifications: 1 executed      1 passed        0 failed        0 skipped
Scenarios:      1 executed      1 passed        0 failed        0 skipped
```

Note the "`?`" character before `Column1`.

This behaviour is because the test code is looking for `Column1`, which does not exist! Instead it finds the zero-width-no-break-space prefixed `Column1`, causing string match to fail.

I am not sure if this can be addressed via Gauge. The way I could get this working was to actually save the file with Encoding: "Utf8 without signature".

**Comment by apoorvam** [comment]
_From @ryanrust on August 22, 2016 20:15_

From what I understand, some text editors silently insert that character into the beginning of a file. While it is possible that you could remove it by changing encoding or by copying the text into another editor, I literally had to build the Gauge source to even see what was happening. I was hoping it would be acceptable (if possible) to strip leading BOM from the stream or however that data is passed into the step implementation. (I haven't had time to dig deeper to see if/where this could even be implemented.)

**Comment by apoorvam** [comment]
_From @sriv on August 25, 2016 3:44_

Fair enough, I think this issue is not specific to Gauge-CSharp, so would be inclined to see if Gauge core would need to sanitize the csv.

**Comment by apoorvam** [comment]
@sguptatw Fix should be available in nightly >= 06.09.2016

[@ryanrust](https://github.com/ryanrust) we are now sanitizing the external csv/text files with encoding say utf-8 with signature. To try this, you can download the latest gauge-nightly from [here](https://bintray.com/gauge/Gauge/Nightly/_latestVersion) and try replicating this issue. 

Let us know how it goes.
